ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/AstSanitizer.java

```diff
@@ -0,0 +1,283 @@
+import io.confluent.ksql.parser.tree.InsertInto;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.SingleColumn;
+import io.confluent.ksql.schema.ksql.ColumnRef;
+import io.confluent.ksql.schema.ksql.FormatOptions;
+import io.confluent.ksql.util.DataSourceExtractor;
+import io.confluent.ksql.util.KsqlConstants;
+import io.confluent.ksql.util.KsqlException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+/**
+ * Validate and clean ASTs generated from externally supplied statements
+ */
+public final class AstSanitizer {
+  private AstSanitizer() {
+  }
+
+  public static AstNode sanitize(final AstNode node, final MetaStore metaStore) {
+    final DataSourceExtractor dataSourceExtractor = new DataSourceExtractor(metaStore);
+    dataSourceExtractor.extractDataSources(node);
+
+    final RewriterPlugin rewriterPlugin = new RewriterPlugin(metaStore, dataSourceExtractor);
+
+    final ExpressionRewriterPlugin expressionRewriterPlugin =
+        new ExpressionRewriterPlugin(metaStore, dataSourceExtractor);
+    final BiFunction<Expression, Void, Expression> expressionRewriter =
+        (e, v) -> ExpressionTreeRewriter.rewriteWith(expressionRewriterPlugin::process, e, v);
+
+    return new StatementRewriter<>(expressionRewriter, rewriterPlugin::process)
+        .rewrite(node, null);
+  }
+
+  private static final class RewriterPlugin extends
+      AstVisitor<Optional<AstNode>, StatementRewriter.Context<Void>> {
+    final MetaStore metaStore;
+    final DataSourceExtractor dataSourceExtractor;
+
+    private int selectItemIndex = 0;
+
+    RewriterPlugin(final MetaStore metaStore, final DataSourceExtractor dataSourceExtractor) {
+      super(Optional.empty());
+      this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+      this.dataSourceExtractor
+          = Objects.requireNonNull(dataSourceExtractor, "dataSourceExtractor");
+    }
+
+    @Override
+    protected Optional<AstNode> visitCreateStreamAsSelect(
+        final CreateStreamAsSelect node,
+        final StatementRewriter.Context<Void> ctx
+    ) {
+      return Optional.of(
+          new CreateStreamAsSelect(
+              node.getLocation(),
+              node.getName(),
+              (Query) ctx.process(node.getQuery()),
+              node.isNotExists(),
+              node.getProperties(),
+              node.getPartitionByColumn()
+          )
+      );
+    }
+
+    @Override
+    protected Optional<AstNode> visitInsertInto(
+        final InsertInto node,
+        final StatementRewriter.Context<Void> ctx
+    ) {
+      final DataSource<?> target = getSource(
+          node.getTarget(),
+          node.getLocation().map(
+              l -> new NodeLocation(
+                  l.getLineNumber(),
+                  l.getColumnNumber() +  "INSERT INTO".length()
+              )
+          )
+      );
+      if (target.getDataSourceType() != DataSourceType.KSTREAM) {
```

Comment:
IMHO, this check does not belong in the parser. The statement should _parse_ fine even if its a table, because the parser should not _know_ it's a table. It is the engine that should reject it, because the engine _knows_ the target is not a stream and _knows_ the statement is only valid for streams.

i.e. I'm proposing we move this check out of the parser all together and into the engine.

(I have a feeling the check already exists in the engine).